### PR TITLE
Hide torch_python symbols

### PR DESCRIPTION
**Pull request by cyyever**
Change symbols in torch_python to invisible by default on platforms other than Apple.

cc @albanD

### CODE REVIEW

**Comment by albanD** [comment]
I'm not sure how to test this :/ 
Can you share the PR that makes that visibility change?

**Comment by cyyever** [comment]
It is #136743

**Review by ezyang — APPROVED** [review]
This is kind of risky but I guess we can give it a try.

**Comment by albanD** [comment]
Curious what is the plan to test this to make sure we didn't miss others? Did you try on a set of third party extensions?

File: torch/CMakeLists.txt [hunk 1]
```diff
@@ -311,10 +311,6 @@ endif()
 add_library(torch_python SHARED ${TORCH_PYTHON_SRCS})
 torch_compile_options(torch_python)  # see cmake/public/utils.cmake
 # target_compile_definitions(torch_python PRIVATE -DEXPORT_AOTI_FUNCTIONS)
-if(NOT WIN32)
-  target_compile_options(torch_python PRIVATE
-      $<$<COMPILE_LANGUAGE:CXX>: -fvisibility=default>)
-endif()
```
[the remark this answers is not in the document]
Comment (cyyever) [reply torch/CMakeLists.txt:317]: torch_compile_options uses private visibility for torch targets. Line 313 is from other PR.

File: torch/csrc/Layout.h [hunk 2]
```diff
@@ -15,7 +15,7 @@ struct THPLayout {
   char name[LAYOUT_NAME_LEN + 1];
 };
 
-extern PyTypeObject THPLayoutType;
+TORCH_PYTHON_API extern PyTypeObject THPLayoutType;
```
[the remark this answers is not in the document]
Comment (cyyever) [reply torch/csrc/Layout.h:18]: I built it locally and fixed all import errors. The 3 APIs were found during that time.

**Comment by cyyever** [comment]
> Curious what is the plan to test this to make sure we didn't miss others? Did you try on a set of third party extensions?

Try it on hugging face libraries.

**Comment by albanD** [comment]
I actually would like to argue to revert this one again.
It broke a bunch of things including some internal use case that had to be fixed in https://github.com/pytorch/pytorch/pull/143380#event-15693208626 and it is also breaking the WIP extension at https://github.com/pytorch/pytorch/tree/main/test/cpp_extensions/open_registration_extension (working on enabling it in CI but failing now).

I am afraid there are a lot more of these APIs that would need to be marked and as you mentioned in the other PR, we don't have a good way right now to make sure we caught all of them.
Also @atalman can you confirm this did NOT make it into the release. If it did, we should definitely remove it!

File: torch/CMakeLists.txt [hunk 3]
```diff
@@ -310,7 +310,7 @@ endif()
 
 add_library(torch_python SHARED ${TORCH_PYTHON_SRCS})
 torch_compile_options(torch_python)  # see cmake/public/utils.cmake
-if(NOT WIN32)
+if(APPLE)
```
Comment (malfet) [inline torch/CMakeLists.txt:313]: Why this check here? Imo symbol visibility on Linux and Mac should be identical, shouldn't it?

Comment (cyyever) [reply torch/CMakeLists.txt:313]: Apple has some hard to find linking errors that lead to test failures

**Comment by suo** [comment]
I think this change makes sense directionally, but making a public symbol private is a breaking change so we should be careful about it. I think it would be good to revert and come up with a plan for how to roll it out.

**Comment by cyyever** [comment]
Another way is to gradually discover those symbols and mark them public.

**Comment by albanD** [comment]
The problem with this approach @cyyever is that users based on the release will have a 3months delay where they're unable to use PyTorch until the next release comes out. Which is really not acceptable.

**Comment by cyyever** [comment]
@albanD Fully understand, seems that it should be reverted.

**Comment by atalman** [comment]
@albanD @malfet I confirm this is nor part of the release 2.6:
https://github.com/pytorch/pytorch/blob/release/2.6/torch/CMakeLists.txt#L313

**Pull request closed without merging**